pkg/controller/machinepool/machinepool_controller.go

```diff
@@ -618,25 +638,51 @@ func (r *ReconcileMachinePool) syncMachineSets(
 					rMS.Spec.Template.Spec.Taints = []corev1.Taint{}
 				}
 				// Make a temporary map of generated MachineSet's taints for easy lookup
-				gTaintsByKey := make(map[string]*corev1.Taint)
+				gTaintsByID := make(map[hivev1.TaintIdentifier]*corev1.Taint)
 				for gIndex, gTaint := range ms.Spec.Template.Spec.Taints {
-					gTaintsByKey[gTaint.Key] = &ms.Spec.Template.Spec.Taints[gIndex]
+					gTaintsByID[identifierForTaint(&gTaint)] = &ms.Spec.Template.Spec.Taints[gIndex]
 				}
-				// Go through the remote MachineSet's taints, replacing overlaps with the generated MachineSet
-				for rIndex, rTaint := range rMS.Spec.Template.Spec.Taints {
-					if gTaint, foundTaint := gTaintsByKey[rTaint.Key]; foundTaint {
-						if gTaint.Value != rTaint.Value || gTaint.Effect != rTaint.Effect {
-							rMS.Spec.Template.Spec.Taints[rIndex] = *gTaint
-							objectModified = true
-						}
-						delete(gTaintsByKey, rTaint.Key)
+				// Build the list of taints for the remote MachineSet from scratch
+				taints := make([]corev1.Taint, 0)
+				// Go through the remote MachineSet's taints, adding/replacing matches from the generated MachineSet
+				for _, rTaint := range rMS.Spec.Template.Spec.Taints {
+					tid := identifierForTaint(&rTaint)
+
+					// If we need to "delete" this one, skip it, but mark modified
+					if taintsToDelete.Has(tid) {
+						objectModified = true
+						continue
 					}
+
+					if gTaint, foundTaint := gTaintsByID[tid]; foundTaint && gTaint.Value != rTaint.Value {
+						// Existing taint is modified
+						rTaint = *gTaint
+						objectModified = true
+					}
+
+					// If we get here, one of the following is true:
+					// - The taint was in the generated MachineSet. Whether it was modified or not, it must be included.
+					// - The taint was in the remote MachineSet but not the generated one. However, it wasn't in our
+					//   "owned" list, i.e. it is "remote only". Preserve it.
+					taints = append(taints, rTaint)
+					// We'll want to include any "extras" from the generated MachineSet
+					delete(gTaintsByID, tid)
 				}
-				// Any remaining taints from the temporary map are new and need to be added
-				for _, gTaint := range gTaintsByKey {
-					rMS.Spec.Template.Spec.Taints = append(rMS.Spec.Template.Spec.Taints, *gTaint)
-					objectModified = true
+				// Anything remaining from the generated MachineSet is new
+				for _, taint := range gTaintsByID {
+					taints = append(taints, *taint)
 				}
+				sort.SliceStable(taints, func(i, j int) bool {
```

Comment:
POSSIBLE BUG: Does taint order matter to k8s? If so, this is not the right thing to do. We should in fact be conveying the ordering from the MachinePool to the extent that is feasible.